Source/Mosa.Compiler.Framework/Stages/LoopRangeTrackerStage.cs

```diff
@@ -109,91 +107,296 @@ private void ProcessNode(Node node, Loop loop, bool is32Bit)
+		if (incrementValue.IsInt32 && (incrementValue.ConstantSigned32 >= short.MaxValue - 1 || incrementValue.ConstantSigned32 <= short.MinValue + 1))
+			return false;
+
+		if (!incrementValue.IsInt32 && (incrementValue.ConstantSigned64 >= short.MaxValue - 1 || incrementValue.ConstantSigned64 <= short.MinValue + 1))
+			return false;
+
+		foreach (var use in incrementVariable.Uses)
 		{
-			if (!(b.Instruction == IR.Branch32 || b.Instruction == IR.Branch64))
+			if (!(use.Instruction == IR.Branch32 || use.Instruction == IR.Branch64))
 				continue;
 
-			// only that are the header or backedge (if only one)
-			if (!(b.Block == loop.Header || (loop.Backedges.Count == 1 && loop.Backedges.Contains(b.Block))))
+			// only analysis the header or backedge (if only one)
+			if (!(use.Block == loop.Header || (loop.Backedges.Count == 1 && loop.Backedges.Contains(use.Block))))
 				continue;
 
-			var x = b.Operand1;
-			var y = b.Operand2;
-			var condition = b.ConditionCode;
-			var target = b.BranchTarget1;
-			var othertarget = target != b.Block.NextBlocks[0]
-				? b.Block.NextBlocks[0]
-				: b.Block.NextBlocks[1];
+			var x = use.Operand1;
+			var y = use.Operand2;
+			var condition = use.ConditionCode;
+			var target = use.BranchTarget1;
+			var othertarget = target != use.Block.NextBlocks[0]
+				? use.Block.NextBlocks[0]
+				: use.Block.NextBlocks[1];
 
-			// form: x (variable) </<=/== y (constant) -> where branch exits the loop
+			// form: x (variable) {>,>=,==} y (constant) -> where branch exits loop
+
+			// change form - on branch
+			if (loop.LoopBlocks.Contains(target))
+			{
+				(condition, target, othertarget) = (condition.GetOpposite(), othertarget, target);
+			}
+
+			// change form - on constant to right
+			if (!y.IsResolvedConstant && (condition == ConditionCode.Equal || condition == ConditionCode.NotEqual))
+			{
+				(x, y) = (y, x);
+			}
 
 			// change form - on condition
-			if (condition is not (ConditionCode.Less
-				or ConditionCode.LessOrEqual
-				or ConditionCode.UnsignedLess
-				or ConditionCode.UnsignedLessOrEqual
-				or ConditionCode.Equal))
+			if (!y.IsResolvedConstant)
 			{
 				(x, y, condition) = (y, x, condition.GetOpposite()); // swap
 			}
 
+			if (condition is not (
+				   ConditionCode.Greater
+				or ConditionCode.GreaterOrEqual
+				or ConditionCode.UnsignedGreater
+				or ConditionCode.UnsignedGreaterOrEqual))
+				continue;
+
+			if (condition == ConditionCode.NotEqual && !incrementValue.IsConstantOne)
+				continue;
+
+			if (condition == ConditionCode.Equal)
+				continue;
+
+			if (x != incrementVariable)
+				continue;
+
+			if (!y.IsResolvedConstant)
+				continue;
+
+			if (incrementValue.IsInt32 && (y.ConstantSigned32 >= short.MaxValue - 1 || y.ConstantSigned32 <= short.MinValue + 1))
+				continue;
+
+			if (!incrementValue.IsInt32 && (y.ConstantSigned64 >= short.MaxValue - 1 || y.ConstantSigned64 <= short.MinValue + 1))
+				continue;
```

Comment:
Shouldn't it be `y.IsInt32` instead of `incrementValue.IsInt32`?
If `y` and `incrementValue` are the same, then no need to perform this check again, since it was already done right before the `foreach` loop started.

Reply:
Technically, it should be the same. However, I will switch to y since that's provide more clarity. Also adding a debug check.